app/src/main/java/com/keylesspalace/tusky/util/LinkHelper.kt

```diff
@@ -70,27 +71,113 @@ fun getDomain(urlString: String?): String {
         styleQuoteSpans(view)
-        getSpans(0, spannableContent.length, URLSpan::class.java).forEach { span ->
+        getSpans(0, endOfContent, URLSpan::class.java).forEach { span ->
+            if (get(getSpanStart(span)) == '#') {
+                inlineHashtagSpanCount += 1
+            }
             setClickableText(span, this, mentions, tags, listener)
         }
-    }
+    }.subSequence(0, endOfContent).trimEnd()
+
     view.movementMethod = NoTrailingSpaceLinkMovementMethod
+
+    val showHashtagBar = (trailingHashtags.isNotEmpty() || inlineHashtagSpanCount != tags?.size)
+    // I don't _love_ setting the visibility here, but the alternative is to duplicate the logic in other places
+    trailingHashtagView?.visible(showHashtagBar)
+
+    if (showHashtagBar) {
+        trailingHashtagView?.apply {
+            text = SpannableStringBuilder().apply {
+                var offset = 0
+                tags?.forEachIndexed { index, tag ->
+                    val text = "#${tag.name}"
+                    append(text)
+                    setSpan(
+                        getCustomSpanForTag(text, tags, URLSpan(tag.url), listener),
+                        offset,
+                        offset + tag.name.length + 1,
+                        0
+                    )
+                    offset += tag.name.length + 1
+                    if (index != tags.lastIndex) {
+                        append(" ")
+                        offset += 1
+                    }
+                }
+            }
+        }
+    }
+}
+
+private val trailingHashtagExpression by unsafeLazy {
+    Pattern.compile("""$WORD_BREAK_EXPRESSION(#$HASHTAG_EXPRESSION$WORD_BREAK_FROM_SPACE_EXPRESSION+)*""", Pattern.CASE_INSENSITIVE)
 }
 
+/**
+ * Find the "trailing" hashtags in spanned content
+ * These are hashtags in lines consisting *only* of hashtags at the end of the post
+ */
+@VisibleForTesting
+internal fun getTrailingHashtags(content: Spanned): Pair<Int, List<HashTag>> {
+    val trailingHashtagStart = content.reversedLineSequence().takeWhile {
+        val line = content.subSequence(it.first, it.second)
+        line.isBlank() || trailingHashtagExpression.matcher(line).matches()
+    }.lastOrNull()?.first ?: return Pair(content.length, emptyList())
+
+    return Pair(
+        trailingHashtagStart,
+        content.getSpans(trailingHashtagStart, content.length, URLSpan::class.java)
+            .filter { content[content.getSpanStart(it)] == '#' } // just in case
+            .map { spanToHashtag(content, it) }
+    )
+}
+
+private val newlineCharacters = arrayOf('\n', '\r').toCharArray()
+
+/**
+ * Get a reversed sequence of start/end line indices (not including the newline characters)
+ * for lazily searching backward by line.
+ * The idea is that we're very likely to early-out, so we want to avoid splitting lots of long messages
+ * in their entirety just to look at the last line or two.
+ * Maybe it's overkill? 🤷
```

Comment:
I think I would go with a simpler implementation and see how it works in practice